src/Symfony/Component/Lock/Store/RedisStore.php

```diff
@@ -0,0 +1,153 @@
+        }
+
+        if ($initialTtl <= 0) {
+            throw new InvalidArgumentException(sprintf('%s() expects a strictly positive TTL. Got %d.', __METHOD__, $initialTtl));
+        }
+
+        $this->redis = $redisClient;
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key)
+    {
+        $script = '
+            if redis.call("GET", KEYS[1]) == ARGV[1] then
+                return redis.call("PEXPIRE", KEYS[1], ARGV[2])
+            else
+                return redis.call("set", KEYS[1], ARGV[1], "NX", "PX", ARGV[2])
+            end
+        ';
+
+        $expire = (int) ceil($this->initialTtl * 1000);
+        if (!$this->evaluate($script, (string) $key, array($this->getToken($key), $expire))) {
+            throw new LockConflictedException();
+        }
+    }
+
+    public function waitAndSave(Key $key)
+    {
+        throw new InvalidArgumentException(sprintf('The store "%s" does not supports blocking locks.', get_class($this)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function putOffExpiration(Key $key, $ttl)
+    {
+        $script = '
+            if redis.call("GET", KEYS[1]) == ARGV[1] then
+                return redis.call("PEXPIRE", KEYS[1], ARGV[2])
+            else
+                return 0
+            end
+        ';
+
+        $expire = (int) ceil($ttl * 1000);
+        if (!$this->evaluate($script, (string) $key, array($this->getToken($key), $expire))) {
+            throw new LockConflictedException();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(Key $key)
+    {
+        $script = '
+            if redis.call("GET", KEYS[1]) == ARGV[1] then
+                return redis.call("DEL", KEYS[1])
+            else
+                return 0
+            end
+        ';
+
+        $this->evaluate($script, (string) $key, array($this->getToken($key)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists(Key $key)
+    {
+        return $this->redis->get((string) $key) === $this->getToken($key);
+    }
+
+    /**
+     * Evaluates a script in the corresponding redis client.
+     *
+     * @param string $script
+     * @param string $resource
+     * @param array  $args
+     *
+     * @return mixed
+     */
+    private function evaluate($script, $resource, array $args)
+    {
+        if ($this->redis instanceof \Redis || $this->redis instanceof \RedisCluster) {
+            return $this->redis->eval($script, array_merge(array($resource), $args), 1);
+        }
+
+        if ($this->redis instanceof \RedisArray) {
+            return $this->redis->_instance($this->redis->_target($resource))->eval($script, array_merge(array($resource), $args), 1);
+        }
+
+        // Have to be a \Predis\Client
+        return call_user_func_array(array($this->redis, 'eval'), array_merge(array($script, 1, $resource), $args));
```

Comment:
I know that constructor allows only some classes and most of them are used above, but wouldn't be more readable to check instance of `\Predis\Client` too? After all `if` statements exception could be thrown that script could not be evaluated.